File: src/Chapter05GraphModellingAndFloodFill/Leetcode785.java
package Chapter05GraphModellingAndFloodFill;

public class Leetcode785 {
    private int[][] graph;
    private int[] colors;
    private boolean[] visited;

    public boolean isBipartite(int[][] graph) {
        this.graph = graph;
        int V = graph.length;
        visited = new boolean[V];
        colors = new int[V];

        for (int v = 0; v < V; v++) {
            if (!visited[v]) {
                if (!dfs(v, 0)) return false;
            }
        }
        return true;
    }

    public boolean dfs(int start, int color) {
        visited[start] = true;
        colors[start] = color;
        for (int v : graph[start]) {
            if (!visited[v]) {
                if (!dfs(v, 1 - color)) {
                    return false;
                }
            } else if (colors[v] == colors[start]) {
                return false;

            }
        }
        return true;
    }
}

File: src/Chapter03GraphDFSApplications/leetcode785.py
from collections import defaultdict


class Solution:
    def isBipartite(self, graph):
        self.graph = graph
        self.visited = [False] * len(graph)
        self.hascycle = False
        self.colors = [-1] * len(graph)
        for i in range(len(graph)):
            if not self.visited[i]:
                if not self.dfs(i, 0):
                    return False

        return True
    def dfs(self, root, color):
        self.visited[root] = True
        self.colors[root] = color
        for w in self.graph[root]:
            if not self.visited[w]:
                if not self.dfs(w, 1 - color):
                    return False
            elif self.colors[w] == color:
                return False
        return True
